datatypes/src/main/java/org/hyperledger/besu/datatypes/AccessEvent.java

```diff
@@ -0,0 +1,307 @@
+/*
+ * Copyright contributors to Hyperledger Besu.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.datatypes;
+
+import java.util.Objects;
+
+import org.apache.tuweni.units.bigints.UInt256;
+
+/**
+ * Manages access events that are done to the stateless AccessWitness. An access mode is passed to
+ * the witness representing the type of access being done on the verkle trie leaf node: read, write
+ * to an existing leaf (reset) or write to a non-existent leaf (set). The access mode is then
+ * evaluated against structures that track previous access events, to avoid double charging of a
+ * warm account or slot. What results is the **new** access event schedule that is then used for gas
+ * charging.
+ *
+ * <p>This class can represent a leaf or a branch access event. Each leaf access event has a
+ * reference to its corresponding branch event, so that accesses to leaves and branches can be
+ * reverted synchronously.
+ *
+ * @param <T> key type for the access event
+ */
+public abstract class AccessEvent<T> {
+
+  private static final long WITNESS_BRANCH_COST = 1900;
+  private static final long WITNESS_CHUNK_COST = 200;
+  private static final long SUBTREE_EDIT_COST = 3000;
+  private static final long CHUNK_EDIT_COST = 500;
+  private static final long CHUNK_FILL_COST = 6200;
+
+  /** Bit mask for NONE */
+  public static final int NONE = 0;
+
+  /** Bit mask for BRANCH READ */
+  public static final int BRANCH_READ = 1;
+
+  /** Bit mask for BRANCH WRITE */
+  public static final int BRANCH_WRITE = 2;
+
+  /** Bit mask for LEAF READ */
+  public static final int LEAF_READ = 4;
+
+  /** Bit mask for LEAF RESET */
+  public static final int LEAF_RESET = 8;
+
+  /** Bit mask for LEAF SET */
+  public static final int LEAF_SET = 16;
+
+  /**
+   * These fields comprise the key to the event, hence only these are used for computing equals and
+   * hashcode. Because an instance of this object can serve as a key and a value, key and values
+   * fields are packed together.
+   */
+  protected final T key;
+
+  private final UInt256 index;
+
+  /** These are the actual values for the event, they are not to be used in equals and hashcode. */
+  private long accessCounter;
+
+  private int accessFlags = NONE;
+
+  private int hash;
```

Comment:
maybe renaming to hashCode ? we are using hash everywhere in besu and I had to check the class to understand what is this hash with int